### ISSUE TITLE
Adding 'Type' to ResultError struct

### ISSUE THREAD

**Issue by cristiangraz** [xeipuuv/gojsonschema#51]
I'm trying to create custom error messages for a JSON API in this format:

```json
"error": {
   "field": "firstName",
   "code": 50001,
   "message": "First name is required"
}
```
The idea is I need to convert validation errors on my JSON schema to a specific error format. It could be an error code, a custom message, or anything else really.

What do you think of adding a 'Type' field to the ResultError struct? The context already has the field name, but other than having a string there is no way to know which error occurred. I put together an example here: https://github.com/cristiangraz/gojsonschema

Essentially you would have values like: enum, missing, less_than, unique, invalid_type, greater_than, less_than, etc to know which error occurred on each field. From there, it gives the developer more options to create customized error messages. Another option would be to change the locales to be these values that you could check against, but I don't see an easy way to overwrite the locale without forking and I think there is value in knowing the error type and still having a human-readable error message.

Open to other suggestions/ideas ways of accomplishing this if there is interest.

**Comment by xeipuuv** [comment]
There is definitely room for improvements with locales in error handling and it is something i have been "struggling" with when writing the library. The main difficulty is i guess to find a way to handle all errors, those that imply one field, those implying two fields etc... Which might result in dealing with different kinds of structures for each error type.

Also this is going to be some kind of archeology work for now :) Making sure all errors are taken in account ( schema creation ones + validation ones ).

I honestly dont know what is the best solution but here are some thoughts:

- We must keep the standard locales modes, forcing the library user to implements all kinds of errors before being able to use it is quite repulsive.

- Allowing the user to override the locales in a Printf fashion can be a good solution. ex : SetErrorLocale( Type, "%field Must be greater than %value" )

- We could provide interfaces and default implementations for all types of errors. So that the user can use the default ones or write/update his own versions. This may also allow them to write their own formatters, a good thing when dealing with JSON APIs for example.

- If there is a single way to represent all kinds of errors, i mean using the same structure, i guess the Type system works. If not, I am afraid the user will have to perform some big ugly switch to handle all types and output the appropriate error.

**Comment by cristiangraz** [comment]
Thanks for the thoughts. There are definitely several different layers of accomplishing this.

The example I put together is the simplest I could think of: Add a new Type field to the struct and identify the error. Then each developer can work with that in their own library to implement their own responses (should they need to) like the example I gave above. The idea here is to just add more context to the ResultError, which is only used for validation - so I did not do anything with schema errors.

The locales currently can't be overwritten because they are constants... I left that as is. 

The user also would not have implement anything to use and this would be backwards compatible. They would just receive an additional Type field on the errors with a string like "invalid_type", "enum", etc. If there are multiple errors, there would be multiple ResultError structs returned, and each of those would have the new Type field so it would still address that use case.

I'm curious to get your thoughts just given the limited scope. What I'm going for is something similar to https://github.com/xeipuuv/gojsonschema/pull/6 which just adds more context to the error without breaking BC.

As a more long-term solution, I think allowing library users to create a locale struct or map[string]string and pass that into the library... otherwise the default locale will be used. This would do away with the locale constants. Each error should contain error details that are specific to that error. For example: invalid type might contain details: { "expected": "string", "given": TYPE_BOOLEAN} which can then be converted into the string replacements %expected% and %given%. GTE might contain ErrorDetails{"min": 10}. The locales would then return a string like: "%field% must be a %expected%, %given% given" and the library will handle the replacements. Each error would be it's own type (i.e. InvalidTypeError and still have a type field for easy comparison).

addError would then be called like this:

```go
result.addError(ErrorTypeInvalid, context, currentNode, ErrorDetails{"expected": currentSubSchema.types.String(), "given": TYPE_STRING})
```
which would return an error struct that looks like this:
```go
InvalidTypeError {
	Type:       "invalid_type",
	Context:     *jsonContext,
	Description: "firstName should be a string, boolean given"
	Value:       value,
	Details:     ErrorDetails{
            "expected": "string"
            "given": "boolean"
        }
}
```
And as a string:

(root).firstName firstName should be a string, boolean given

Schema errors could be handled the same way, but the call to errors.New() would just be a string that is generated in the same way as validation errors.
